### ISSUE TITLE
[RFC] ECHO terminal training recipe

### ISSUE THREAD

**Issue by kylemontgomery1** [verl-project/verl-recipe#132]
### Summary
This RFC proposes an ECHO training recipe for verl. ECHO augments GRPO with an on-policy next-token cross-entropy objective over environment-observation tokens. The recipe implements this objective through a generic auxiliary-token loss interface, allowing agent loops to select arbitrary response tokens for an additional cross-entropy term while retaining the standard policy-gradient objective over model-generated actions.

### Motivation
Existing verl agent loops can train on rewards assigned to generated actions, but they do not directly optimize prediction of environment observations. The ECHO objective is a small, opt-in extension that supports the main experiment from the paper without introducing a broader world-model abstraction or changing behavior for existing users.

### Proposal
The implementation has two parts:
- Native auxiliary-token loss support: add an optional `aux_token_loss_coeff` actor setting, propagate an optional `aux_token_loss_mask` from `AgentLoopOutput` into the training batch, and add masked token cross-entropy to the PPO policy loss when the coefficient is nonzero. The default coefficient is zero, and the auxiliary loss uses the actor's configured `loss_agg_mode` and global batch normalization information. The implementation spans four files:
  - `verl/experimental/agent_loop/agent_loop.py`: add `aux_token_loss_mask` to `AgentLoopOutput`, convert it to a tensor, and pad and batch it.
  - `verl/trainer/config/actor/actor.yaml and verl/workers/config/actor.py`: add `aux_token_loss_coeff` with a default of zero.
  - `verl/workers/utils/losses.py`: when the coefficient is positive, compute masked token cross-entropy and add `aux_token_loss_coeff × aux_token_loss` to the PPO policy loss.
- ECHO recipe: provide data preparation, the reference's Qwen3 XML chat template, bash and done tools, a terminal agent loop, Modal sandbox integration, and a training launcher for the main Qwen3-8B experiment. The agent loop selects environment-observation tokens through aux_token_loss_mask.

### Data
The full training set was not released. The recipe therefore uses [obiwan96/endless-terminals](https://huggingface.co/datasets/obiwan96/endless-terminals), which is a public subset. We sample 100 tasks for validation and use the remaining 2,392 tasks for training. To match the reference and train for 500 steps requires > 3 epochs.

### Minor differences from reference
- Use the verl XML tool parser, which has less granular malformed-tool-call warnings than the reference implementation.
- Normalize the auxiliary loss over selected environment tokens rather than the reference full-observation-token count. The latter requires passing an additional count through from the agent loop to the loss, in addition to `aux_token_loss_mask`.

### Validation
- Validated on both v0 and v1 trainer implementations:

<img height="300" alt="Image" src="https://github.com/user-attachments/assets/b2b71c80-4217-4ff0-9d14-f17061fd216a" />
<img height="300" alt="Image" src="https://github.com/user-attachments/assets/47608957-0eb3-4aad-964a-2621987fa751" />

**[ByteDance Internal Only]**
verl: https://code.byted.org/data/open_verl/tree/kyle/echo-repro
verl-recipe: https://code.byted.org/kylemontgomery/verl-recipes/tree/kyle/echo-repro

### References
Paper: https://arxiv.org/abs/2605.24517
Code: https://github.com/microsoft/echo-rl
